From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RuscalWorld <me@ruscalworld.ru>
Date: Tue, 28 Dec 2021 22:00:04 +0300
Subject: [PATCH] Masquerade protected commands

Add option in utopia.yml that allows to replace "no permission" message with "unknown command" message, so players will not be able to determine if the protected command exist on server.

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 1f45b04f3cd3645b0fb3b994b27609cfaa04662b..1284d331c441519853c051fe559022981f3052b2 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -4,7 +4,6 @@ import com.google.common.base.Charsets;
 import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Iterators;
 import com.google.common.collect.Lists;
 import com.google.common.collect.MapMaker;
@@ -13,13 +12,12 @@ import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.tree.CommandNode;
 import com.mojang.brigadier.tree.LiteralCommandNode;
-import com.mojang.serialization.DynamicOps;
 import com.mojang.serialization.Lifecycle;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufOutputStream;
 import io.netty.buffer.Unpooled;
 import io.papermc.paper.logging.SysoutCatcher;
-import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap;
+
 import java.awt.image.BufferedImage;
 import java.io.File;
 import java.io.FileInputStream;
@@ -29,7 +27,6 @@ import java.io.InputStreamReader;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Base64;
 import java.util.Collections;
 import java.util.HashSet;
@@ -58,7 +55,6 @@ import net.minecraft.commands.arguments.EntityArgument;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.MappedRegistry;
 import net.minecraft.core.Registry;
-import net.minecraft.resources.RegistryReadOps;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.ConsoleInput;
@@ -2788,6 +2784,8 @@ public final class CraftServer implements Server {
 
     @Override
     public String getPermissionMessage() {
+        if (ru.bortexel.utopia.UtopiaConfig.masqueradeProtectedCommands)
+            return org.spigotmc.SpigotConfig.unknownCommandMessage;
         return com.destroystokyo.paper.PaperConfig.noPermissionMessage;
     }
 
diff --git a/src/main/java/ru/bortexel/utopia/UtopiaConfig.java b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
index 1307e823962c63f5644b999ce0f3c80800b2ba24..2aae07d298de57c670dd695b54730bc8bb6771f2 100644
--- a/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
+++ b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
@@ -125,4 +125,10 @@ public class UtopiaConfig {
     private static void loadControlledProperty(ControlledProperty property) {
         controlledProperties.put(property.getName(), property);
     }
+
+    public static String STUFF_ROOT = "stuff";
+    public static boolean masqueradeProtectedCommands = false;
+    private static void configStuff() {
+        masqueradeProtectedCommands = getBoolean(STUFF_ROOT + ".masquerade-protected-commands", false);
+    }
 }
